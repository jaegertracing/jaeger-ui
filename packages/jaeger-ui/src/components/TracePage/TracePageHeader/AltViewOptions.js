// @flow

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Button, Dropdown, Icon, Menu } from 'antd';
import { Link } from 'react-router-dom';

import { trackAltViewOpen } from './TracePageHeader.track';
import prefixUrl from '../../../utils/prefix-url';

type Props = {
  onTraceGraphViewClicked: () => void,
  traceGraphView: boolean,
  traceID: string,
};

export default function AltViewOptions(props: Props) {
  const { onTraceGraphViewClicked, traceGraphView, traceID } = props;
  const menu = (
    <Menu>
      <Menu.Item>
        <a onClick={onTraceGraphViewClicked}>{traceGraphView ? 'Trace Timeline' : 'Trace Graph'}</a>
      </Menu.Item>
      <Menu.Item>
        <Link
          to={prefixUrl(`/api/traces/${traceID}?prettyPrint=true`)}
          rel="noopener noreferrer"
          target="_blank"
          onClick={trackAltViewOpen}
        >
          Trace JSON
        </Link>
      </Menu.Item>
      <Menu.Item>
        <Link
          to={prefixUrl(`/api/traces/${traceID}?raw=true&prettyPrint=true`)}
          rel="noopener noreferrer"
          target="_blank"
          onClick={trackAltViewOpen}
        >
          Trace JSON (unadjusted)
        </Link>
      </Menu.Item>
    </Menu>
  );
  return (
    <Dropdown overlay={menu}>
      <Button className="ub-mr2" onClick={onTraceGraphViewClicked}>
        {traceGraphView ? 'Trace Graph' : 'Trace Timeline'} <Icon type="down" />
      </Button>
    </Dropdown>
  );
}
