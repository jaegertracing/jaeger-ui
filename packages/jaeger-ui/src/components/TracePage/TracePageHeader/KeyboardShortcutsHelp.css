/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.KeyboardShortcutsHelp--cta {
  font-size: 1.2rem;
  margin: 0 -7px;
}

.KeyboardShortcutsHelp--table {
  background-color: #fafafa;
  max-height: 500px;
  overflow: auto;
}

.KeyboardShortcutsHelp--oddRow {
  background-color: #f6f6f6;
}

.KeyboardShortcutsHelp--table kbd {
  background: #f0f0f0;
  border-bottom: 1px solid #bbb;
  border-radius: 3px;
  border: 1px solid #d4d4d4;
  color: #000;
  font-family: monospace;
  padding: 0.25em 0.3em;
}
