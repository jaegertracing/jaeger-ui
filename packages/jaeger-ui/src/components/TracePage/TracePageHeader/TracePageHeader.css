/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.TracePageHeader > :first-child {
  border-bottom: 1px solid #e8e8e8;
}

.TracePageHeader > :nth-child(2) {
  background-color: #eee;
  border-bottom: 1px solid #e4e4e4;
}

.TracePageHeader > :last-child {
  background-color: #f8f8f8;
  border-bottom: 1px solid #ccc;
}

/* boost the specificity for cases with only one row -- bg should be white */
.TracePageHeader > .TracePageHeader--titleRow {
  align-items: center;
  background-color: #fff;
  display: flex;
}

.TracePageHeader--back {
  align-items: center;
  align-self: stretch;
  background-color: #fafafa;
  border-bottom: 1px solid #ddd;
  border-right: 1px solid #ddd;
  color: inherit;
  display: flex;
  font-size: 1.4rem;
  padding: 0 1rem;
  margin-bottom: -1px;
}

.TracePageHeader--back:hover {
  background-color: #f0f0f0;
  border-color: #ccc;
}

.TracePageHeader--titleLink {
  align-items: center;
  color: var(--tx-color-title);
  display: flex;
  flex: 1;
}

.TracePageHeader--titleLink:hover * {
  text-decoration: underline;
}

.TracePageHeader--titleLink:hover > *,
.TracePageHeader--titleLink:hover small {
  text-decoration: none;
}

.TracePageHeader--detailToggle {
  font-size: 2.5rem;
  transition: transform 0.07s ease-out;
}

.TracePageHeader--detailToggle.is-expanded {
  transform: rotate(90deg);
}

.TracePageHeader--title {
  color: inherit;
  flex: 1;
  font-size: 1.7em;
  line-height: 1em;
  margin: 0 0 0 0.5em;
  padding: 0.5em 0;
}

.TracePageHeader--title.is-collapsible {
  margin-left: 0;
}

.TracePageHeader--overviewItems {
  border-bottom: 1px solid #e4e4e4;
  padding: 0.25rem 0.5rem;
}

.TracePageHeader--archiveIcon {
  font-size: 1.78em;
  margin-right: 0.15em;
}
