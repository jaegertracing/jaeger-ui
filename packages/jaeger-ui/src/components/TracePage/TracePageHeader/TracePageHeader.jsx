// This is a partial implementation focusing on the Archive button component

import React from 'react';
import { Button, Tooltip, message } from 'antd';
import { DownOutlined } from '@ant-design/icons';
import { connect } from 'react-redux';

import * as jaegerApiActions from '../../../actions/jaeger-api';
import { getArchivedEnabled } from '../../../utils/config/get-config';

export class TracePageHeaderFn extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      archiveInProgress: false,
    };
  }

  archiveTrace = () => {
    const { id, archiveTrace } = this.props;
    this.setState({ archiveInProgress: true });
    
    // Check if using in-memory storage
    const isInMemoryStorage = this.props.storage && this.props.storage.type === 'memory';
    
    if (isInMemoryStorage) {
      // Show warning toast for in-memory storage
      message.warning(
        'Warning: You are archiving to in-memory storage. Archived traces will be lost when Jaeger restarts.',
        5
      );
    }
    
    archiveTrace(id)
      .then(() => {
        this.setState({ archiveInProgress: false });
        message.success('This trace has been archived');
      })
      .catch(error => {
        this.setState({ archiveInProgress: false });
        message.error(`Unable to archive trace: ${error.message}`);
      });
  };

  render() {
    const { archiveEnabled, ...otherProps } = this.props;
    const { archiveInProgress } = this.state;
    
    return (
      <div className="TracePageHeader">
        {/* Other header components */}
        
        {archiveEnabled && (
          <Tooltip
            title="Archive this trace for long-term storage"
            placement="bottom"
          >
            <Button
              className="ArchiveButton"
              disabled={archiveInProgress}
              onClick={this.archiveTrace}
              loading={archiveInProgress}
            >
              Archive Trace <DownOutlined />
            </Button>
          </Tooltip>
        )}
        
        {/* Other UI elements */}
      </div>
    );
  }
}

// mapStateToProps function
function mapStateToProps(state, ownProps) {
  const { config, router } = state;
  const { capabilities } = router.location.state || {};
  
  // Use getArchivedEnabled to determine if archive button should be shown
  // This properly handles in-memory storage case
  const archiveEnabled = getArchivedEnabled(config, capabilities);
  
  return {
    ...ownProps,
    archiveEnabled,
    storage: config.storage,
  };
}

// mapDispatchToProps function
function mapDispatchToProps(dispatch) {
  return {
    archiveTrace: id => dispatch(jaegerApiActions.archiveTrace(id)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(TracePageHeaderFn);
