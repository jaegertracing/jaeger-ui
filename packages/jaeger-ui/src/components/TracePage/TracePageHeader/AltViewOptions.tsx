// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Dropdown, Icon, Menu } from 'antd';
import { Link } from 'react-router-dom';
import './AltViewOptions.css';

import {
  trackGanttView,
  trackGraphView,
  trackStatisticsView,
  trackJsonView,
  trackRawJsonView,
} from './TracePageHeader.track';
import prefixUrl from '../../../utils/prefix-url';
import { ETraceViewType } from '../types';

type Props = {
  onTraceViewChange: (viewType: ETraceViewType) => void;
  traceID: string;
  viewType: ETraceViewType;
};

const MENU_ITEMS = [
  {
    viewType: ETraceViewType.TraceTimelineViewer,
    label: 'Trace Timeline',
  },
  {
    viewType: ETraceViewType.TraceGraph,
    label: 'Trace Graph',
  },
  {
    viewType: ETraceViewType.TraceStatistics,
    label: 'Trace Statistics',
  },
];

export default function AltViewOptions(props: Props) {
  const { onTraceViewChange, viewType, traceID } = props;

  const handleSelectView = (item: ETraceViewType) => {
    if (item === ETraceViewType.TraceTimelineViewer) {
      trackGanttView();
    } else if (item === ETraceViewType.TraceGraph) {
      trackGraphView();
    } else if (item === ETraceViewType.TraceStatistics) {
      trackStatisticsView();
    }
    onTraceViewChange(item);
  };

  const menu = (
    <Menu>
      {MENU_ITEMS.filter(item => item.viewType !== viewType).map(item => (
        <Menu.Item key={item.viewType}>
          <a onClick={() => handleSelectView(item.viewType)} role="button">
            {item.label}
          </a>
        </Menu.Item>
      ))}
      <Menu.Item>
        <Link
          to={prefixUrl(`/api/traces/${traceID}?prettyPrint=true`)}
          rel="noopener noreferrer"
          target="_blank"
          onClick={trackJsonView}
        >
          Trace JSON
        </Link>
      </Menu.Item>
      <Menu.Item>
        <Link
          to={prefixUrl(`/api/traces/${traceID}?raw=true&prettyPrint=true`)}
          rel="noopener noreferrer"
          target="_blank"
          onClick={trackRawJsonView}
        >
          Trace JSON (unadjusted)
        </Link>
      </Menu.Item>
    </Menu>
  );

  const currentItem = MENU_ITEMS.find(item => item.viewType === viewType);
  const dropdownText = currentItem ? currentItem.label : 'Alternate Views';
  return (
    <Dropdown overlay={menu}>
      <div className="AltViewOptions">
        {`${dropdownText} `}
        <Icon type="down" />
      </div>
    </Dropdown>
  );
}
