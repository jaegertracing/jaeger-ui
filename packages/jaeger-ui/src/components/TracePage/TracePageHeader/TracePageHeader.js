// @flow

// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Button, Input } from 'antd';
import _maxBy from 'lodash/maxBy';
import _values from 'lodash/values';
import IoAndroidArrowBack from 'react-icons/lib/io/android-arrow-back';
import IoIosFilingOutline from 'react-icons/lib/io/ios-filing-outline';
import MdKeyboardArrowRight from 'react-icons/lib/md/keyboard-arrow-right';
import { Link } from 'react-router-dom';

import AltViewOptions from './AltViewOptions';
import KeyboardShortcutsHelp from './KeyboardShortcutsHelp';
import SpanGraph from './SpanGraph';
import TracePageSearchBar from './TracePageSearchBar';
import LabeledList from '../../common/LabeledList';
import NewWindowIcon from '../../common/NewWindowIcon';
import TraceName from '../../common/TraceName';
import { getTraceName } from '../../../model/trace-viewer';
import { formatDatetime, formatDuration } from '../../../utils/date';

import type { ViewRange, ViewRangeTimeUpdate } from '../types';
import type { Trace } from '../../../types/trace';

import './TracePageHeader.css';

type TracePageHeaderEmbedProps = {
  canCollapse: boolean,
  clearSearch: () => void,
  forwardedRef: { current: Input | null },
  hideMap: boolean,
  hideSummary: boolean,
  linkToStandalone: string,
  nextResult: () => void,
  onArchiveClicked: () => void,
  onSlimViewClicked: () => void,
  onTraceGraphViewClicked: () => void,
  prevResult: () => void,
  resultCount: number,
  showArchiveButton: boolean,
  showShortcutsHelp: boolean,
  showStandaloneLink: boolean,
  showViewOptions: boolean,
  slimView: boolean,
  textFilter: string,
  toSearch: string | null,
  trace: Trace,
  traceGraphView: boolean,
  updateNextViewRangeTime: ViewRangeTimeUpdate => void,
  updateTextFilter: string => void,
  updateViewRangeTime: (number, number) => void,
  viewRange: ViewRange,
};

export const HEADER_ITEMS = [
  {
    key: 'timestamp',
    label: 'Trace Start',
    renderer: (trace: Trace) => formatDatetime(trace.startTime),
  },
  {
    key: 'duration',
    label: 'Duration',
    renderer: (trace: Trace) => formatDuration(trace.duration),
  },
  {
    key: 'service-count',
    label: 'Services',
    renderer: (trace: Trace) => new Set(_values(trace.processes).map(p => p.serviceName)).size,
  },
  {
    key: 'depth',
    label: 'Depth',
    renderer: (trace: Trace) => _maxBy(trace.spans, 'depth').depth + 1,
  },
  {
    key: 'span-count',
    label: 'Total Spans',
    renderer: (trace: Trace) => trace.spans.length,
  },
];

export function TracePageHeaderFn(props: TracePageHeaderEmbedProps) {
  const {
    canCollapse,
    clearSearch,
    forwardedRef,
    hideMap,
    hideSummary,
    linkToStandalone,
    nextResult,
    onArchiveClicked,
    onSlimViewClicked,
    onTraceGraphViewClicked,
    prevResult,
    resultCount,
    showArchiveButton,
    showShortcutsHelp,
    showStandaloneLink,
    showViewOptions,
    slimView,
    textFilter,
    toSearch,
    trace,
    traceGraphView,
    updateNextViewRangeTime,
    updateTextFilter,
    updateViewRangeTime,
    viewRange,
  } = props;

  if (!trace) {
    return null;
  }

  const summaryItems =
    !hideSummary &&
    !slimView &&
    HEADER_ITEMS.map(item => {
      const { renderer, ...rest } = item;
      return { ...rest, value: renderer(trace) };
    });

  const title = (
    <h1 className={`TracePageHeader--title ${canCollapse ? 'is-collapsible' : ''}`}>
      <TraceName traceName={getTraceName(trace.spans)} />{' '}
      <small className="u-tx-muted">{trace.traceID.slice(0, 7)}</small>
    </h1>
  );

  return (
    <header className="TracePageHeader">
      <div className="TracePageHeader--titleRow">
        {toSearch && (
          <Link className="TracePageHeader--back" to={toSearch}>
            <IoAndroidArrowBack />
          </Link>
        )}
        {canCollapse ? (
          <a
            className="TracePageHeader--titleLink"
            onClick={onSlimViewClicked}
            role="switch"
            aria-checked={!slimView}
          >
            <MdKeyboardArrowRight
              className={`TracePageHeader--detailToggle ${!slimView ? 'is-expanded' : ''}`}
            />
            {title}
          </a>
        ) : (
          title
        )}
        {showShortcutsHelp && <KeyboardShortcutsHelp className="ub-mr2" />}
        <TracePageSearchBar
          updateTextFilter={updateTextFilter}
          textFilter={textFilter}
          prevResult={prevResult}
          nextResult={nextResult}
          clearSearch={clearSearch}
          resultCount={resultCount}
          ref={forwardedRef}
        />

        {showViewOptions && (
          <AltViewOptions
            onTraceGraphViewClicked={onTraceGraphViewClicked}
            traceGraphView={traceGraphView}
            traceID={trace.traceID}
          />
        )}
        {showArchiveButton && (
          <Button className="ub-mr2 ub-flex ub-items-center" onClick={onArchiveClicked}>
            <IoIosFilingOutline className="TracePageHeader--archiveIcon" />
            Archive Trace
          </Button>
        )}
        {showStandaloneLink && (
          <Link
            className="u-tx-inherit ub-nowrap ub-mx2"
            to={linkToStandalone}
            target="_blank"
            rel="noopener noreferrer"
          >
            <NewWindowIcon />
          </Link>
        )}
      </div>
      {summaryItems && <LabeledList className="TracePageHeader--overviewItems" items={summaryItems} />}
      {!hideMap &&
        !slimView && (
          <SpanGraph
            trace={trace}
            viewRange={viewRange}
            updateNextViewRangeTime={updateNextViewRangeTime}
            updateViewRangeTime={updateViewRangeTime}
          />
        )}
    </header>
  );
}

// ghetto fabulous cast because the 16.3 API is not in flow yet
// https://github.com/facebook/flow/issues/6103
export default (React: any).forwardRef((props, ref) => <TracePageHeaderFn {...props} forwardedRef={ref} />);
