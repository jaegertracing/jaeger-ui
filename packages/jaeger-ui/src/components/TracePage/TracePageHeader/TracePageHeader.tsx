// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Button, Input } from 'antd';
import _get from 'lodash/get';
import _maxBy from 'lodash/maxBy';
import _values from 'lodash/values';
import IoAndroidArrowBack from 'react-icons/lib/io/android-arrow-back';
import IoIosFilingOutline from 'react-icons/lib/io/ios-filing-outline';
import MdKeyboardArrowRight from 'react-icons/lib/md/keyboard-arrow-right';
import { Link } from 'react-router-dom';

import AltViewOptions from './AltViewOptions';
import KeyboardShortcutsHelp from './KeyboardShortcutsHelp';
import SpanGraph from './SpanGraph';
import TracePageSearchBar from './TracePageSearchBar';
import { TUpdateViewRangeTimeFunction, IViewRange, ViewRangeTimeUpdate, ETraceViewType } from '../types';
import LabeledList from '../../common/LabeledList';
import NewWindowIcon from '../../common/NewWindowIcon';
import TraceName from '../../common/TraceName';
import { getTraceName } from '../../../model/trace-viewer';
import { TNil } from '../../../types';
import { Trace } from '../../../types/trace';
import { formatDatetime, formatDuration } from '../../../utils/date';
import { getTraceLinks } from '../../../model/link-patterns';

import './TracePageHeader.css';
import ExternalLinks from '../../common/ExternalLinks';

type TracePageHeaderEmbedProps = {
  canCollapse: boolean;
  clearSearch: () => void;
  focusUiFindMatches: () => void;
  hideMap: boolean;
  hideSummary: boolean;
  linkToStandalone: string;
  nextResult: () => void;
  onArchiveClicked: () => void;
  onSlimViewClicked: () => void;
  onTraceViewChange: (viewType: ETraceViewType) => void;
  prevResult: () => void;
  resultCount: number;
  showArchiveButton: boolean;
  showShortcutsHelp: boolean;
  showStandaloneLink: boolean;
  showViewOptions: boolean;
  slimView: boolean;
  textFilter: string | TNil;
  toSearch: string | null;
  trace: Trace;
  viewType: ETraceViewType;
  updateNextViewRangeTime: (update: ViewRangeTimeUpdate) => void;
  updateViewRangeTime: TUpdateViewRangeTimeFunction;
  viewRange: IViewRange;
};

export const HEADER_ITEMS = [
  {
    key: 'timestamp',
    label: 'Trace Start',
    renderer: (trace: Trace) => {
      const dateStr = formatDatetime(trace.startTime);
      const match = dateStr.match(/^(.+)(\.\d+)$/);
      return match ? (
        <span className="TracePageHeader--overviewItem--value">
          {match[1]}
          <span className="TracePageHeader--overviewItem--valueDetail">{match[2]}</span>
        </span>
      ) : (
        dateStr
      );
    },
  },
  {
    key: 'duration',
    label: 'Duration',
    renderer: (trace: Trace) => formatDuration(trace.duration),
  },
  {
    key: 'service-count',
    label: 'Services',
    renderer: (trace: Trace) => new Set(_values(trace.processes).map(p => p.serviceName)).size,
  },
  {
    key: 'depth',
    label: 'Depth',
    renderer: (trace: Trace) => _get(_maxBy(trace.spans, 'depth'), 'depth', 0) + 1,
  },
  {
    key: 'span-count',
    label: 'Total Spans',
    renderer: (trace: Trace) => trace.spans.length,
  },
];

export function TracePageHeaderFn(props: TracePageHeaderEmbedProps & { forwardedRef: React.Ref<Input> }) {
  const {
    canCollapse,
    clearSearch,
    focusUiFindMatches,
    forwardedRef,
    hideMap,
    hideSummary,
    linkToStandalone,
    nextResult,
    onArchiveClicked,
    onSlimViewClicked,
    onTraceViewChange,
    prevResult,
    resultCount,
    showArchiveButton,
    showShortcutsHelp,
    showStandaloneLink,
    showViewOptions,
    slimView,
    textFilter,
    toSearch,
    trace,
    viewType,
    updateNextViewRangeTime,
    updateViewRangeTime,
    viewRange,
  } = props;

  if (!trace) {
    return null;
  }

  const links = getTraceLinks(trace);

  const summaryItems =
    !hideSummary &&
    !slimView &&
    HEADER_ITEMS.map(item => {
      const { renderer, ...rest } = item;
      return { ...rest, value: renderer(trace) };
    });

  const title = (
    <h1 className={`TracePageHeader--title ${canCollapse ? 'is-collapsible' : ''}`}>
      <TraceName traceName={getTraceName(trace.spans)} />{' '}
      <small className="u-tx-muted">{trace.traceID.slice(0, 7)}</small>
    </h1>
  );

  return (
    <header className="TracePageHeader">
      <div className="TracePageHeader--titleRow">
        {toSearch && (
          <Link className="TracePageHeader--back" to={toSearch}>
            <IoAndroidArrowBack />
          </Link>
        )}
        {links && links.length > 0 && <ExternalLinks links={links} />}
        {canCollapse ? (
          <a
            className="TracePageHeader--titleLink"
            onClick={onSlimViewClicked}
            role="switch"
            aria-checked={!slimView}
          >
            <MdKeyboardArrowRight
              className={`TracePageHeader--detailToggle ${!slimView ? 'is-expanded' : ''}`}
            />
            {title}
          </a>
        ) : (
          title
        )}
        <TracePageSearchBar
          clearSearch={clearSearch}
          focusUiFindMatches={focusUiFindMatches}
          nextResult={nextResult}
          prevResult={prevResult}
          ref={forwardedRef}
          resultCount={resultCount}
          textFilter={textFilter}
          navigable={viewType === ETraceViewType.TraceTimelineViewer}
        />
        {showShortcutsHelp && <KeyboardShortcutsHelp className="ub-m2" />}
        {showViewOptions && (
          <AltViewOptions onTraceViewChange={onTraceViewChange} traceID={trace.traceID} viewType={viewType} />
        )}
        {showArchiveButton && (
          <Button className="ub-mr2 ub-flex ub-items-center" htmlType="button" onClick={onArchiveClicked}>
            <IoIosFilingOutline className="TracePageHeader--archiveIcon" />
            Archive Trace
          </Button>
        )}
        {showStandaloneLink && (
          <Link
            className="u-tx-inherit ub-nowrap ub-mx2"
            to={linkToStandalone}
            target="_blank"
            rel="noopener noreferrer"
          >
            <NewWindowIcon isLarge />
          </Link>
        )}
      </div>
      {summaryItems && <LabeledList className="TracePageHeader--overviewItems" items={summaryItems} />}
      {!hideMap && !slimView && (
        <SpanGraph
          trace={trace}
          viewRange={viewRange}
          updateNextViewRangeTime={updateNextViewRangeTime}
          updateViewRangeTime={updateViewRangeTime}
        />
      )}
    </header>
  );
}

export default React.forwardRef((props: TracePageHeaderEmbedProps, ref: React.Ref<Input>) => (
  <TracePageHeaderFn {...props} forwardedRef={ref} />
));
