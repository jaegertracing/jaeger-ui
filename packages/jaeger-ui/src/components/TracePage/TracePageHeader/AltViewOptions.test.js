// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';
import { Dropdown } from 'antd';
import { Link } from 'react-router-dom';
import AltViewOptions from './AltViewOptions';
import * as track from './TracePageHeader.track';
import { ETraceViewType } from '../types';

describe('AltViewOptions', () => {
  let trackGanttView;
  let trackGraphView;
  let trackJsonView;
  let trackRawJsonView;
  let trackStatisticsView;

  let wrapper;
  const getLink = text => {
    const menu = shallow(wrapper.find(Dropdown).prop('overlay'));
    const links = menu.find(Link);
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i);
      if (link.children().text() === text) return link;
    }
    const links2 = menu.find('a');
    for (let i = 0; i < links.length; i++) {
      const link = links2.at(i);
      if (link.children().text() === text) return link;
    }
    throw new Error(`Could not find "${text}"`);
  };

  const props = {
    viewType: ETraceViewType.TraceTimelineViewer,
    traceID: 'test trace ID',
    onTraceViewChange: jest.fn(),
  };

  beforeAll(() => {
    trackGanttView = jest.spyOn(track, 'trackGanttView');
    trackGraphView = jest.spyOn(track, 'trackGraphView');
    trackJsonView = jest.spyOn(track, 'trackJsonView');
    trackRawJsonView = jest.spyOn(track, 'trackRawJsonView');
    trackStatisticsView = jest.spyOn(track, 'trackStatisticsView');
  });

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AltViewOptions {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('tracks viewing JSONs', () => {
    expect(trackJsonView).not.toHaveBeenCalled();
    getLink('Trace JSON').simulate('click');
    expect(trackJsonView).toHaveBeenCalledTimes(1);

    expect(trackRawJsonView).not.toHaveBeenCalled();
    getLink('Trace JSON (unadjusted)').simulate('click');
    expect(trackRawJsonView).toHaveBeenCalledTimes(1);

    expect(trackJsonView).toHaveBeenCalledTimes(1);
    expect(trackGanttView).not.toHaveBeenCalled();
    expect(trackGraphView).not.toHaveBeenCalled();
  });

  it('track dropdown menu', () => {
    const viewInteractions = [
      {
        link: 'Trace Graph',
        trackFn: trackGraphView,
        onTraceViewChangeArg: ETraceViewType.TraceGraph,
      },
      {
        link: 'Trace Statistics',
        trackFn: trackStatisticsView,
        onTraceViewChangeArg: ETraceViewType.TraceStatisticsView,
        propViewType: ETraceViewType.TraceGraph,
      },
      {
        link: 'Trace Timeline',
        trackFn: trackGanttView,
        onTraceViewChangeArg: ETraceViewType.TraceTimelineViewer,
        propViewType: ETraceViewType.TraceStatisticsView,
      },
    ];

    viewInteractions.forEach(({ link, trackFn, propViewType }, i) => {
      if (propViewType) {
        wrapper.setProps({ viewType: propViewType });
      }
      expect(props.onTraceViewChange).toHaveBeenCalledTimes(i);
      expect(trackFn).not.toHaveBeenCalled();

      getLink(link).simulate('click');
      expect(props.onTraceViewChange).toHaveBeenCalledTimes(i + 1);
      viewInteractions.forEach(({ trackFn: fn }, j) => {
        expect(fn).toHaveBeenCalledTimes(j <= i ? 1 : 0);
      });
    });
  });
});
