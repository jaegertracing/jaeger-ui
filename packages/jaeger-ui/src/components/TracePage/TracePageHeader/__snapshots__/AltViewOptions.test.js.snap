// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AltViewOptions renders correctly 1`] = `
<Dropdown
  mouseEnterDelay={0.15}
  mouseLeaveDelay={0.1}
  overlay={
    <Menu
      className=""
      focusable={false}
      prefixCls="ant-menu"
      theme="light"
    >
      <MenuItem>
        <a
          onClick={[Function]}
          role="button"
        >
          Trace Graph
        </a>
      </MenuItem>
      <MenuItem>
        <a
          onClick={[Function]}
          role="button"
        >
          Trace Statistics
        </a>
      </MenuItem>
      <MenuItem>
        <Link
          onClick={[MockFunction]}
          rel="noopener noreferrer"
          replace={false}
          target="_blank"
          to="/api/traces/test trace ID?prettyPrint=true"
        >
          Trace JSON
        </Link>
      </MenuItem>
      <MenuItem>
        <Link
          onClick={[MockFunction]}
          rel="noopener noreferrer"
          replace={false}
          target="_blank"
          to="/api/traces/test trace ID?raw=true&prettyPrint=true"
        >
          Trace JSON (unadjusted)
        </Link>
      </MenuItem>
    </Menu>
  }
  placement="bottomLeft"
  prefixCls="ant-dropdown"
>
  <div
    className="AltViewOptions"
  >
    Trace Timeline 
    <Icon
      type="down"
    />
  </div>
</Dropdown>
`;
